File: javascript/src/validation.ts
import { JustDeployValidationError } from './errors.js';

export function pathSegment(value: string, label: string): string {
  if (typeof value !== 'string' || value.length === 0) {
    throw new JustDeployValidationError(`${label} must be a non-empty string.`);
  }
  return encodeURIComponent(value);
}

export function pageQuery(options: { limit?: number; cursor?: number }, maxLimit: number): string {
  const params = new URLSearchParams();
  if (options.limit !== undefined) {
    if (!Number.isInteger(options.limit) || options.limit <= 0 || options.limit > maxLimit) {
      throw new JustDeployValidationError(`limit must be an integer between 1 and ${maxLimit}.`);
    }
    params.set('limit', String(options.limit));
  }
  if (options.cursor !== undefined) {
    if (!Number.isInteger(options.cursor) || options.cursor <= 0) {
      throw new JustDeployValidationError('cursor must be a positive integer.');
    }
    params.set('cursor', String(options.cursor));
  }
  const query = params.toString();
  return query ? `?${query}` : '';
}

File: python/src/justdeploy/_validation.py
from urllib.parse import quote, urlencode

from .errors import JustDeployValidationError


def path_segment(value: str, label: str) -> str:
    if not isinstance(value, str) or not value:
        raise JustDeployValidationError(f"{label} must be a non-empty string.")
    return quote(value, safe="")


def page_query(*, limit: int | None, cursor: int | None, max_limit: int) -> str:
    values: dict[str, str] = {}
    if limit is not None:
        if isinstance(limit, bool) or not isinstance(limit, int) or not 1 <= limit <= max_limit:
            raise JustDeployValidationError(f"limit must be an integer between 1 and {max_limit}.")
        values["limit"] = str(limit)
    if cursor is not None:
        if isinstance(cursor, bool) or not isinstance(cursor, int) or cursor <= 0:
            raise JustDeployValidationError("cursor must be a positive integer.")
        values["cursor"] = str(cursor)
    return f"?{urlencode(values)}" if values else ""
